README.md
Devel: [![Bioconductor devel build Status](http://bioconductor.org/shields/build/devel/bioc/pRolocGUI.svg)](http://bioconductor.org/packages/devel/bioc/html/pRolocGUI.html)
 - Release: [![Bioconductor release build Status](http://bioconductor.org/shields/build/release/bioc/pRolocGUI.svg)](http://bioconductor.org/packages/release/bioc/html/pRolocGUI.html)

Exploring and visualising spatial proteomics data
=================================================

## Introduction

The
[`pRolocGUI`](http://www.bioconductor.org/packages/devel/bioc/html/pRolocGUI.html)
package is an interactive interface to explore and visualise
experimental mass spectrometry-based spatial proteomics data. It
relies on the [`shiny`](http://shiny.rstudio.com/) framework for
interactive visualisation, the
[`MSnbase`](http://www.bioconductor.org/packages/release/bioc/html/MSnbase.html)
package to handle data and metadata and the
[`pRoloc`](http://www.bioconductor.org/packages/release/bioc/html/pRoloc.html)
software for spatial proteomics specific data matters. Example spatial
data is available in the
[`pRolocdata`](http://bioconductor.org/packages/release/data/experiment/html/pRolocdata.html)
experiment package.


The `pRoloc` suite set of software are distributed as part of the
`R`/[Bioconductor](http://bioconductor.org/) project and are developed
at the [Computational Proteomics Unit](http://cpu.sysbiol.cam.ac.uk/)
at the
[Cambridge Centre for Proteomics](http://proteomics.bio.cam.ac.uk)
labs, at the University of Cambridge.

This document describes the installation of the software, followed by
a basic quick start guide for using `pRolocGUI` to search and
visualise the organelle map. Please refer to the respective
documentation and vignettes for full details about the software.

If you use these open-source software for your research, please cite:

>  Gatto L, Breckels LM, Wieczorek S, Burger T, Lilley KS.
>  Mass-spectrometry-based spatial proteomics data analysis using
>  pRoloc and pRolocdata. Bioinformatics. 2014 May 1;30(9):1322-4.
>  doi:10.1093/bioinformatics/btu013. Epub 2014 Jan 11. PMID:24413670;
>  PMCID:PMC3998135.

>  Breckels LM, Gatto L, Christoforou A, Groen AJ, Lilley KS, Trotter
>  MW. The effect of organelle discovery upon sub-cellular protein
>  localisation. J Proteomics. 2013 Mar 21. doi:pii:
>  S1874-3919(13)00094-8. 10.1016/j.jprot.2013.02.019. PMID:23523639.

>  Gatto L., Breckels L.M., Burger T, Nightingale D.J.H., Groen A.J.,
>  Campbell C., Mulvey C.M., Christoforou A., Ferro M., Lilley K.S. 'A
>  foundation for reliable spatial proteomics data analysis' Mol Cell
>  Proteomics. 2014 May 20.

## Installation

`pRolocGUI` is written in the [`R`](http://www.r-project.org/)
programming language. Before installing the software you need to
download `R` and (optionally) `RStudio`.

1) Download the latest `R` release for your operating system from the
[R website](http://www.r-project.org/) and install it.

2) Optional, but
recommended. [Download](http://www.rstudio.com/products/rstudio/download/)
and install the RStudio IDE. `RStudio` provides a good code editor and
excellent integration with the `R` terminal.

3) Start `R` or `RStudio`.

4) Install the Bioconductor packages `pRoloc`, `pRolocdata` and
`pRolocGUI`:

`pRolocGUI` requires `R >= 3.1.1` and Bioconductor version `>= 3.0`.
In an `R` console, type

```
if (!requireNamespace("BiocManager", quietly=TRUE))
	install.packages("BiocManager")
BiocManager::install(c("pRoloc", "pRolocdata", "pRolocGUI"))
```

#### Development version

The development code on github can also be installed using
`BiocManager::install` (or `install_github`). New pre-release features
might not be documented or thoroughly tested and could substantially
change prior to release. Use at your own risks.

```
BiocManager::install("ComputationalProteomicsUnit/pRolocGUI")
```

## Getting started

Before using a package's functionality, it needs to be loaded:

```
library("pRolocGUI")
```

We first load data from
[Christoforou et al 2016](http://www.nature.com/ncomms/2016/160112/ncomms9992/full/ncomms9992.html)
distributed in the `pRolocdata` package:

```
library("pRolocdata")
data(hyperLOPIT2015)
```

There are 3 different visualisation applications currently
available: `main`, `classification` and `compare`.
These apps are launched using the `pRolocVis` function and
passing `object`, which is an `MSnSet` containing the data
one wishes to interrogate. One may also specify which app
they wish to use by using the `app` argument, see `?pRolocVis`
for more details. The default app that is loaded if
`app` is not specified is the `main` application:

```
pRolocVis(hyperLOPIT2015)
```

![Screenshot - PCA](https://raw.githubusercontent.com/ComputationalProteomicsUnit/pRolocGUI/master/vignettes/figures/SS_PCA1.jpg)

The graphical interface is described in details in the package
vignette that is included in the package itself (get it by typing
`vignette("pRolocGUI")` in `R`), available by clicking the `?` once
the interface is loaded or can be
[consulted online](http://bioconductor.org/packages/devel/bioc/vignettes/pRolocGUI/inst/doc/pRolocGUI.html).

## More resources

### Support

* The Bioconductor [support forum](https://support.bioconductor.org/)
* Open a `pRolocGUI` GitHub
  [issue](https://github.com/ComputationalProteomicsUnit/pRolocGUI/issues)
  (requires a free GitHub account).

### Videos (new videos will appear shortly for the new apps)

0. [An introduction to Bioconductor](https://www.youtube.com/watch?v=dg6NvmMVQ3I)
1. [A brief introduction to `pRolocGUI`](http://youtu.be/zXtiiAoB_vM)
2. [Downloading and install `R`](http://youtu.be/qHMUnZnrkdA)
3. [Using RStudio](http://youtu.be/vlSbKf0OK3g)
4. [Installing the `pRolocGUI` interface](http://youtu.be/_VmAZF1g_O4)
5. [Starting `pRolocGUI`](http://youtu.be/24gntTd133w) - This tutorial is for the older legacy applications. New videos will appear shortly for the new applications.
6. [Using `pRolocGUI` to explore and visualise experimental spatial proteomics data](http://youtu.be/cG3PEQ-uWhM) - This tutorial is for the older legacy applications. New videos will appear shortly for the new applications.

Tutorial [playlist](https://www.youtube.com/watch?v=qHMUnZnrkdA&list=PLvIXxpatSLA2loV5Srs2VBpJIYUlVJ4ow).

### General resources

- [Teaching material](http://lgatto.github.io/TeachingMaterial/)
- R and Bioconductor for proteomics
  [web page](http://lgatto.github.io/RforProteomics/) and
  [package](http://www.bioconductor.org/packages/release/data/experiment/html/RforProteomics.html).


NEWS
CHANGES IN VERSION 1.17
-----------------------
- New verion for Bioc devel 3.9

CHANGES IN VERSION 1.16
-----------------------
- New verion for Bioc release 3.8

CHANGES IN VERSION 1.15.2
-------------------------
- Update NEWS and pkgdown site

CHANGES IN VERSION 1.15.1
-------------------------
- Replace BiocInstaller biocLite mentions with BiocManager

CHANGES IN VERSION 1.15.0
-------------------------
- New version for Bioc devel

CHANGES IN VERSION 1.14.0
-------------------------
- New version for Bioc release

CHANGES IN VERSION 1.13.2
-------------------------
- Catch unnecessary idDT error produced on exit <2018-04-23 Mon>

CHANGES IN VERSION 1.13.1
-------------------------
- Update pRolocVis documentation do show how to use precomputed
  coordinates (see issue #92) <2018-03-16 Fri>

CHANGES IN VERSION 1.12.0
-------------------------
- Biocondutor release 3.6

CHANGES IN VERSION 1.11.3
-------------------------
- Rename the main app pca app <2017-08-21 Mon>

CHANGES IN VERSION 1.11.2
-------------------------
- Fix links in vignette to point to new html pRoloc vignettes
  <2017-05-30 Tue>

CHANGES IN VERSION 1.11.1
-------------------------
- Avoid computing dimensionality reduction at every reactive
  rendering, assuring that other, slower methods, in particular t-SNE,
  can be used <2017-05-20 Sat>


CHANGES IN VERSION 1.11.0
-------------------------
- New version for Bioc devel 3.6


CHANGES IN VERSION 1.10.0
-------------------------
- New version for Bioc release 3.5


CHANGES IN VERSION 1.9.5
------------------------

- Fix broken documentation files <2017-01-18 Wed>
- New aggregation app added <2017-01-22 Wed>

CHANGES IN VERSION 1.9.4
------------------------

- fixed remap=FALSE bug in compare app <2017-01-12 Thu>
- Added mirrorX and mirrorY to the compare app <2017-01-12 Thu>

CHANGES IN VERSION 1.9.3
------------------------

- Update vignette use latest BiocStyle::html_document2() with
   floating table-f content <2016-12-30 Fri>
- Change to NEWS.md <2016-12-30 Fri>
- mirrorX and mirrorY are now ignored in pRolocVis - see issue #84
  <2017-01-11 Wed>

CHANGES IN VERSION 1.9.2
------------------------

- Remove accidental merging left-over <2016-12-13 Tue>

CHANGES IN VERSION 1.9.1
------------------------

- Remove accidental call to browser

CHANGES IN VERSION 1.9.0
------------------------

- Bioc devel 3.5


CHANGES IN VERSION 1.8.0
------------------------

- Bioc release 3.4


CHANGES IN VERSION 1.7.5
------------------------

- Various bug fixes <2016-10-07 Fri>

CHANGES IN VERSION 1.7.4
------------------------

- Add DT version dependency (see issue #71) <2016-08-09 Tue>

CHANGES IN VERSION 1.7.3
------------------------

- Fixed bug with matrix-f markers in fcol. Now if-ne-f the
   columns in the feature data is a matrix, it is converted to
   a vector using mrkMatToVec. This helps clarity and reduces
   wasted table space-f 1's and 0's. <2016-07-29 Fri>
- Fixed bug in pRolocVis_compare application. Same issue as
   above with passing matrix as a fcol which manifested as an
   extra issue with zooming. Now fixed. <2016-07-29 Fri>


CHANGES IN VERSION 1.5.7
------------------------

- Fixed bug in pRolocVis in converting GO ids <2016-05-26 Thu>

CHANGES IN VERSION 1.5.6
------------------------

- Removing plotMat2D app (closes issue #69) <2016-03-11 Fri>
- add package startup msg <2016-03-11 Fri>
- instruct users to install latest version from github

CHANGES IN VERSION 1.5.5
------------------------

- Depend-n DT >= 0.1.40

CHANGES IN VERSION 1.5.4
------------------------

- replace getLisacol by getStockcol (which are now Lisa's colours;
   since pRoloc version 1.11.9) <2016-02-16 Tue>

CHANGES IN VERSION 1.5.3
------------------------

- update refs in vignette <2016-02-09 Tue>

CHANGES IN VERSION 1.5.2
------------------------

- Updated pca app <2016-01-11 Mon>
- Updated vignette <2016-01-12 Tue>
- Fixed bugs, pca app renamed main app, removed
   profiles app <2016-01-14 Thu>
- new compare app <2016-01-30 Sat>
- updated vignette <2016-02-03 Wed>

CHANGES IN VERSION 1.5.1
------------------------

- New shiny apps <2015-10-12 Mon>
- New vignette <2015-10-29 Thu>
- Fixed bugs and updated examples in classify app
   <2015-11-09 Mon>

CHANGES IN VERSION 1.5.0
------------------------

- Bioc devel 3.3


CHANGES IN VERSION 1.4.0
------------------------

- Bioc release 3.2


CHANGES IN VERSION 1.3.2
------------------------

- Fix check warnings and errors <2015-07-20 Mon>

CHANGES IN VERSION 1.3.1
------------------------

- new plotMat2D function <2015-05-20 Wed>
- Fix query search in pRolocVis, contributed by pierremj
   <2015-05-27 Wed>
- pRolocVis has new method arg <2015-05-29 Fri>

CHANGES IN VERSION 1.3.0
------------------------

- Devel for Bioc 3.2


CHANGES IN VERSION 1.2.0
------------------------

- Release for Bioc 3.1


CHANGES IN VERSION 1.1.5
------------------------

- updating failing unit test [2015-04-03 Fri]

CHANGES IN VERSION 1.1.4
------------------------

- fix R_HOME error [2015-02-26 Thu]

CHANGES IN VERSION 1.1.3
------------------------

- don't require GNU make [2015-02-11 Wed]

CHANGES IN VERSION 1.1.2
------------------------

- Fix bug with FoIs and multiple data sets (reported by Harriet
   Parsons) [2015-02-06 Fri]

CHANGES IN VERSION 1.1.1
------------------------

- Update README with uptodate installation instructions
   [2014-10-14 Tue]
- handling and filtering missing value in input data [2014-10-21 Tue]


CHANGES IN VERSION 1.0.0
------------------------

- Release for Bioc 3.0 [2014-10-14 Tue]


CHANGES IN VERSION 0.99.12
--------------------------

- added Video tag in DESCRIPTION [2014-10-07 Tue]

CHANGES IN VERSION 0.99.11
--------------------------

- Add screenshot to README [2014-09-05 Fri]
- fix bug when features-f fois are not present
- move data tab to the end

CHANGES IN VERSION 0.99.10
--------------------------

- Updated README [2014-09-04 Thu]
- Remove-ld R code file [2014-09-04]


R/pRolocVis2.R
## TODO:
## - profiles selection and highlighting

## Shiny: spinning loading wheel on top of plot while plot is recalculating
## https://gist.github.com/daattali/edd7c20cd09f484b7f32

## Very slow with bigger data (fusion data), server side table flaky
## and warnings about size

## Possibly automate mrkVecToMat?

## References
## http://shiny.rstudio.com/articles/plot-interaction-advanced.html
## https://gallery.shinyapps.io/095-plot-interaction-advanced/
## https://gallery.shinyapps.io/105-plot-interaction-zoom/
## https://gallery.shinyapps.io/106-plot-interaction-exclude/
## https://github.com/rstudio/shiny-examples


## http://shiny.rstudio.com/articles/selecting-rows-of-data.html
## http://shiny.rstudio.com/articles/plot-interaction-advanced.html

##' pRoloc interactive visualisation
##' 
##' @title Visualise your pRoloc data
##' @param object An instance of class \code{MSnSet}.
##' @param fcol The name of the markers matrix (default is
##' \code{"Markers"}). Can be missing if \code{foi} is available.
##' @param foi A \code{\link{FeaturesOfInterest}} or a
##' \code{\link{FoICollection}}, that will be available for display.
##' @param fig.height Height of the figure. Default is \code{"600px"}.
##' @param fig.width Width of the figure. Default is \code{"600px"}.
##' @param legend.width Width of the legend. Default is \code{"100\%"}.
##' @param legend.cex Character expansion for the vignette
##' labels. Default is 1.
##' @param nchar Maximum number of characters if the markers class
##' names, before their names are truncated. Default is 10.
##' @param all If there are more than 10 clusters, only the first
##' three are discplayed on start-up, unless \code{all} is set to
##' \code{TRUE}. Default is \code{FALSE}.
##' @param ... Additional parameters that can be used to choose the
##' dimentionality reduction method, as defined in
##' \code{\link{plot2D}}.
##' @author Laurent Gatto
##' @examples
##' library("pRoloc")
##' library("pRolocdata")
##' data(dunkley2006)
##' ## markers matrix ecoding
##' dunkley2006 <- mrkVecToMat(dunkley2006)
##' ## order the fractions 
##' dunkley2006 <- dunkley2006[, order(dunkley2006$fraction)]
##' if (interactive())
##'   pRolocGUI:::pRolocVis2(dunkley2006)
pRolocVis2 <- function(object, fcol,
                       foi,
                       fig.height = "600px",
                       fig.width = "100%",
                       legend.width = "100%",
                       legend.cex = 1,
                       nchar = 15,
                       all = FALSE,
                       ...) {
    if (!inherits(object, "MSnSet"))
        stop("The input must be of class MSnSet")
    if (missing(foi) & missing(fcol)) 
        fcol <- "Markers"
    if (!missing(fcol)) {
        if (!fcol %in% fvarLabels(object))
            stop("fcol missing in fData")
        if (!isMrkMat(object, fcol))
            stop("Markers must be encoded as a matrix. See ?markers for details.")
        pmarkers <- fData(object)[, fcol]
    }
    ## Setting features to be displayed
    if (!missing(foi)) {
        if (inherits(foi, "FeaturesOfInterest"))
            foi <- FoICollection(list(foi))
        foimarkers <- as(foi, "matrix")
        if (exists("pmarkers", inherits = FALSE)) {
            pmarkers <- merge(pmarkers, foimarkers,
                              by = 0, all.x = TRUE)
            rownames(pmarkers) <- pmarkers[, "Row.names"]
            pmarkers <- pmarkers[featureNames(object), -1]            
        } else pmarkers <- foimarkers
    }
    if (length(grep("GO:", colnames(pmarkers))) > 0) {
        cn <- pRoloc::flipGoTermId(colnames(pmarkers))
        if (all(!is.na(cn))) {
            names(cn) <- NULL
            colnames(pmarkers) <- cn
        }
    }
    ## Marker colours
    cols <- getStockcol()
    if (length(cols) < ncol(pmarkers)) {
        message("Too many features for available colours. Some colours will be duplicated.")
        n <- ncol(pmarkers) %/% length(cols)
        cols <- rep(cols, n + 1)
    }
    ## Remove fcol from fData(object)    
    fData(object) <- fData(object)[, -grep(fcol, fvarLabels(object))]
    ## There can't be more than 12 columns in the DT table
    if ((nfd <- length(fvarLabels(object))) > 12) {
        message("There can't be more than 12 feature variables. Using 6 first and last.")
        fData(object) <- fData(object)[, c(1:6, (nfd-5):nfd)]
    }
    ## a hyphen in a pmarkers name breaks the app?!?
    colnames(pmarkers) <- gsub("-", "", colnames(pmarkers))
    ## Shorten markers names if too long
    cn <- sapply(colnames(pmarkers),
                 function(x) {
                     if (nchar(x) > nchar) {
                         x <- strsplit(x, "")[[1]]
                         x <- paste(x[1:nchar], collapse = "")
                         x <- sub(" +$", "", x)
                         x <- paste0(x, "...")
                     }
                     return(x)
                 })    
    names(cn) <- NULL
    colnames(pmarkers) <- cn

    ## If there are too many marker sets, better
    ## to display few and let the user choose
    pmsel <- TRUE
    if (!all & ncol(pmarkers) > 10)
        pmsel <- 1:3

    ## data to be displayed
    pcas <- pRoloc::plot2D(object, fcol = NULL, plot = FALSE, ...)
    profs <- exprs(object)

    ## all feautres are displayed on start
    feats <- 1:nrow(object)
    
    ## Build shiny app
    ui <- fluidPage(
        sidebarLayout(
            sidebarPanel(
                selectizeInput("markers", "Markers",
                               choices = colnames(pmarkers),
                               multiple = TRUE,
                               selected = colnames(pmarkers)[pmsel]),
                sliderInput("trans", "Transparancy",
                            min = 0,  max = 1, value = 0.5),
                checkboxInput("checkbox", label = "Show labels", value = TRUE),
                width = 2),
            mainPanel(
                tabsetPanel(type = "tabs",
                            tabPanel("PCA",
                                     fluidRow(
                                         column(10, 
                                                plotOutput("pca",
                                                           height = fig.height,
                                                           width = fig.width,
                                                           click = "pcaClick",
                                                           dblclick = "pcaDblclick",
                                                           hover = hoverOpts(
                                                               id = "pcaHover"
                                                               ),
                                                           brush = brushOpts(
                                                               id = "pcaBrush",
                                                               resetOnNew = TRUE)),
                                                offset = 0),
                                         column(2, 
                                                plotOutput("legend",
                                                           height = fig.height,
                                                           width = legend.width))
                                         )
                                     ),
                            tabPanel("Profiles",
                                     fluidRow(
                                         column(9, offset = 1,
                                                plotOutput("profile",
                                                           height = fig.height,
                                                           width = fig.width)))
                                     ),
                            ## feature data table is always visible
                            fluidRow(
                                column(12,
                                       column(ncol(fData(object)), 
                                              DT::dataTableOutput("fDataTable"))))
                            )
                )
            ))
    
    server <-
        function(input, output, session) {
            ranges <- reactiveValues(x = NULL, y = NULL)
            ## Get coords for proteins according to selectized marker class(es)
            pcaMrkSel <- reactive({
                lapply(input$markers,
                       function(z) pcas[which(pmarkers[, z] == 1), ])
            })
            profMrkSel <- reactive({
                lapply(input$markers,
                       function(z) profs[which(pmarkers[, z] == 1), ])
            })

            ## Update colour transparacy according to slider input
            myCols <- reactive({
                scales::alpha(cols,
                              input$trans)[sapply(input$markers, function(z) 
                                  which(colnames(pmarkers) == z))]
            })
            ## PCA plot
            output$pca <- renderPlot({
                par(mar = c(4, 4, 0, 0))
                par(oma = c(1, 0, 0, 0))
                plot(pcas,
                     col = getUnknowncol(),
                     pch = 21, cex = 1,
                     xlim = ranges$x,
                     ylim = ranges$y)
                usr <<- par("usr")
                for (i in 1:length(input$markers)) 
                    points(pcaMrkSel()[[i]], pch = 16, cex = 1.4, col = myCols()[i])
                ## FIXME this does not work when brushed/subset of points selected
                s <- feats[input$fDataTable_rows_selected]
                if (length(s)) {
                    points(pcas[s, , drop = FALSE],
                           pch = 19, cex = 2, col = "#00000060")
                    if (input$checkbox)
                        text(pcas[s, 1], pcas[s, 2],
                             rownames(pcas)[s],
                             pos = 1)
                }
            })
            ## Protein profile
            output$profile <- renderPlot({
                par(mar = c(5.1, 4.1, 1, 1))
                matplot(t(profs[feats, ]),
                        col = getUnknowncol(),
                        lty = 1,
                        type = "l")
                ## for (i in 1:length(input$markers)) 
                ##     matlines(t(profMrkSel()[[i]]),
                ##              col = myCols()[i],
                ##              lty = 1,
                ##              lwd = 1.5)
                s <- feats[input$fDataTable_rows_selected]
                if (length(s))
                    matlines(t(profs[s, , drop = FALSE]),
                             col = "black",
                             lty = 1,
                             lwd = 2)
            })                        
            ## Freature data table
            output$fDataTable <- DT::renderDataTable({
                if (is.null(input$pcaBrush)) {
                    i <- try(pcas[, 1] >= usr[1] & pcas[, 1] <= usr[2])
                    j <- try(pcas[, 2] >= usr[3] & pcas[, 2] <= usr[4])
                } else {
                    i <- pcas[, 1] >= input$pcaBrush$xmin & pcas[, 1] <= input$pcaBrush$xmax
                    j <- pcas[, 2] >= input$pcaBrush$ymin & pcas[, 2] <= input$pcaBrush$ymax
                }
                feats <<- which(i & j)
                DT::datatable(fData(object)[i & j, ],
                              rownames = TRUE,
                              options = list(searchHighlight = TRUE))
                              ## filter = 'top')
            })
            ## When a double-click happens, check if there's a brush on the plot.
            ## If so, zoom to the brush bounds; if not, reset the zoom.
            observeEvent(input$pcaDblclick, {
                             brush <- input$pcaBrush
                             if (!is.null(brush)) {
                                 ranges$x <- c(brush$xmin, brush$xmax)
                                 ranges$y <- c(brush$ymin, brush$ymax)
                             } else {
                                 ranges$x <- NULL
                                 ranges$y <- NULL
                             }                                         
                         })
            ## Output legend
            output$legend <- renderPlot({
                par(mar = rep(1, 4))
                par(oma = c(1, 0, 0, 0))
                plot(0, type = "n",
                     xaxt = "n", yaxt = "n",
                     xlab = "", ylab = "",
                     bty = "n")
                legend("topleft",
                       input$markers,
                       col = myCols(),
                       ncol = 1, bty = "n",
                       pch = 16,
                       cex = legend.cex)
            })
        }
    app <- list(ui = ui, server = server)
    runApp(app)
}


## feats
##  features to display on PCA plot
##  profiles to diplay on matplot
##  features to show in DT::datatable

## feats[input$fDataTable_rows_selected]
##  features to highlight
##  feature selected in DT::datatable
